### ISSUE TITLE
elpy-shell-kill-all does not find the *Python* buffer

### ISSUE THREAD

**Issue by bgoodr** [jorgenschaefer/elpy#1147]
Hi,

I think I have found a bug and a possible fix:

elpy-shell-kill-all is bound to C-c C-k.  I expected it to kill the
`*Python*` buffer but it does not find it. Looking at the copy of elpy
I have (elpy-20170613.1339) shows:

    (defun elpy-shell-kill-all (&optional kill-buffers ask-for-each-one)
      "Kill all active python shells.
    
    If KILL-BUFFERS is non-nil, also kill the associated buffers.
    If ASK-FOR-EACH-ONE is non-nil, ask before killing each python process.
    "
      (interactive)
      (let ((python-buffer-list ()))
        ;; Get active python shell buffers and kill inactive ones (if asked)
        (loop for buffer being the buffers do
              (when (and (buffer-name buffer)
                         (string-match "^\*Python\\\[.*\\]\*$" (buffer-name buffer))
                         )
                (if (get-buffer-process buffer)
                    (push buffer python-buffer-list)
                  (when kill-buffers
                    (kill-buffer buffer)))))
        (cond
         ;; Ask for each buffers and kill
         ((and python-buffer-list ask-for-each-one)
          (loop for buffer in python-buffer-list do
                (when (y-or-n-p (format "Kill %s ?" buffer))
                  (delete-process buffer)
                  (when kill-buffers
                    (kill-buffer buffer)))))
         ;; Ask and kill every buffers
         (python-buffer-list
          (if (y-or-n-p (format "Kill %s python shells ?" (length python-buffer-list)))
              (loop for buffer in python-buffer-list do
                    (delete-process buffer)
                    (when kill-buffers
                      (kill-buffer buffer)))))
         ;; No shell to close
         (t
          (message "No python shell to close")))))

The expression:

    (string-match "^\*Python\\\[.*\\]\*$" (buffer-name buffer))

Does not seem to match the buffer named `*Python*`.

Here is my fixed version that uses the handy rx macro:

    (defun elpy-shell-kill-all (&optional kill-buffers ask-for-each-one)
      "Kill all active python shells.
    
    If KILL-BUFFERS is non-nil, also kill the associated buffers.
    If ASK-FOR-EACH-ONE is non-nil, ask before killing each python process.
    "
      (interactive)
      (let ((python-buffer-list ()))
        ;; Get active python shell buffers and kill inactive ones (if asked)
        (loop for buffer being the buffers do
              (when (and (buffer-name buffer)
                         (string-match (rx bol "*Python" (opt "[" (* (not (any "]"))) "]") "*" eol) (buffer-name buffer))
                         )
                (if (get-buffer-process buffer)
                    (push buffer python-buffer-list)
                  (when kill-buffers
                    (kill-buffer buffer)))))
        (cond
         ;; Ask for each buffers and kill
         ((and python-buffer-list ask-for-each-one)
          (loop for buffer in python-buffer-list do
                (when (y-or-n-p (format "Kill %s ?" buffer))
                  (delete-process buffer)
                  (when kill-buffers
                    (kill-buffer buffer)))))
         ;; Ask and kill every buffers
         (python-buffer-list
          (if (y-or-n-p (format "Kill %s python shells ?" (length python-buffer-list)))
              (loop for buffer in python-buffer-list do
                    (delete-process buffer)
                    (when kill-buffers
                      (kill-buffer buffer)))))
         ;; No shell to close
         (t
          (message "No python shell to close")))))

My guess is that it was intended to match multiple python interpreter
buffers, but missed the case where there is only one `*Python*`
buffer. So I made the regular expression match both cases.

Thanks,
Brent

**Comment by jorgenschaefer** [comment]
Nice! Could you make a PR for this? :-)

**Comment by bgoodr** [comment]
Thanks. I created PR at https://github.com/jorgenschaefer/elpy/pull/1149 accordingly.

FYI: Another possibly related area of confusion for me was the meaning of "kill" in that function. "Kill" to me means not only kill the process, but also kill the Emacs buffer. I noticed that the function does kill the process associated with the `*Python*` buffers, but the `*Python*` buffer itself is not killed (deleted from Emacs buffer list).  This may or may not be a feature: This function does cause a message to be added to the end of the `*Python*` buffer, but when you restart the interactive shell, that very same `*Python*` buffer is reused.  Anyhow I did not mess with that behavior so as to just limit this PR to just the original bug I noted.
